manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,431 @@
+
+
+class RemoteWorldState(WorldState):
+    def __init__(self, url: str):
+        actual = urlparse(url)
+        expected = ParseResult(scheme="", netloc="", path=url, params="", query="", fragment="")
+        if actual != expected:
+            raise EthereumError("URL must be of the form 'IP:PORT': " + url)
+        self._url = url
+
+    def _web3(self) -> Web3:
+        # sam.moelius: Force WebsocketProvider.__init__ to call _get_threaded_loop.  The existing
+        # "threaded loop" could be leftover from a fork, in which case it will not work.
+        Web3.WebsocketProvider._loop = None
+        web3 = Web3(Web3.WebsocketProvider(URI("ws://" + self._url)))
+        blocknumber = web3.eth.blockNumber
+        endpoint = _endpoints.get(self._url)
+        if endpoint is None:
+            endpoint = Endpoint(blocknumber, False)
+            _endpoints[self._url] = endpoint
+            logger.info("Connected to %s (blocknumber = %d)", self._url, blocknumber)
+        if endpoint.blocknumber != blocknumber:
+            if not endpoint.warned:
+                logger.warning(
+                    "%s blocknumber has changed: %d != %d---someone is using the endpoint besides us",
+                    self._url,
+                    endpoint.blocknumber,
+                    blocknumber,
+                )
+                endpoint.warned = True
+        return web3
+
+    def accounts(self) -> Set[int]:
+        raise NotImplementedError
+
+    def get_nonce(self, address: int) -> int:
+        return self._web3().eth.getTransactionCount(_web3_address(address))
+
+    def get_balance(self, address: int) -> int:
+        return self._web3().eth.getBalance(_web3_address(address))
+
+    def has_storage(self, address: int) -> bool:
+        raise NotImplementedError
+
+    def get_storage(self, address) -> Storage:
+        raise NotImplementedError
+
+    def get_storage_data(self, address: int, offset: Union[int, BitVec]) -> int:
+        if not isinstance(offset, int):
+            raise NotImplementedError
+        return int.from_bytes(self._web3().eth.getStorageAt(_web3_address(address), offset), "big")
+
+    def get_code(self, address: int) -> bytes:
+        return self._web3().eth.getCode(_web3_address(address))
+
+    def get_blocknumber(self) -> int:
+        return self._web3().eth.blockNumber
+
+    def get_timestamp(self) -> int:
+        return self._web3().eth.getBlock("latest")["timestamp"]
+
+    def get_difficulty(self) -> int:
+        return self._web3().eth.getBlock("latest")["difficulty"]
+
+    def get_gaslimit(self) -> int:
+        return self._web3().eth.getBlock("latest")["gasLimit"]
+
+    def get_coinbase(self) -> int:
+        return int(self._web3().eth.coinbase)
+
+
+####################################################################################################
+
+
+# sam.moelius: If we decide to cache results returned from a RemoteWorldState, then they should NOT
```

Comment:
Something like this?
```
class OverlayWorldState(WorldState):
    """
    If we decide to cache results returned from a RemoteWorldState, then they should NOT be cached
    within an overlay.  The reason is that this could affect the results of subsequent operations.
    Consider a call to get_storage_data followed by a call to has_storage.  If nothing was written
    to storage within the overlay, then the call to has_storage will throw an exception.  But if the
    result of the call to get_storage_data was cached in the overlay, then no exception would be
    thrown.
    """
    ...
```